common/src/main/java/com/wynntils/features/user/ItemLockFeature.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright © Wynntils 2022.
+ * This file is released under AGPLv3. See LICENSE for full license details.
+ */
+package com.wynntils.features.user;
+
+import com.google.common.reflect.TypeToken;
+import com.mojang.blaze3d.vertex.PoseStack;
+import com.wynntils.core.config.Config;
+import com.wynntils.core.config.ConfigManager;
+import com.wynntils.core.config.TypeOverride;
+import com.wynntils.core.features.UserFeature;
+import com.wynntils.core.features.properties.FeatureInfo;
+import com.wynntils.core.features.properties.RegisterKeyBind;
+import com.wynntils.core.keybinds.KeyHolder;
+import com.wynntils.mc.event.ContainerClickEvent;
+import com.wynntils.mc.event.ContainerRenderEvent;
+import com.wynntils.mc.event.DropHeldItemEvent;
+import com.wynntils.mc.event.InventoryKeyPressEvent;
+import com.wynntils.mc.render.RenderUtils;
+import com.wynntils.mc.render.Texture;
+import com.wynntils.mc.utils.McUtils;
+import com.wynntils.wc.model.Character;
+import com.wynntils.wc.utils.WynnUtils;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
+import net.minecraft.world.entity.player.Inventory;
+import net.minecraft.world.inventory.ClickType;
+import net.minecraft.world.inventory.Slot;
+import net.minecraft.world.item.ItemStack;
+import net.minecraftforge.eventbus.api.EventPriority;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import org.lwjgl.glfw.GLFW;
+
+@FeatureInfo(category = "IWillRemoveThisWhenIFix#151")
+public class ItemLockFeature extends UserFeature {
+
+    @RegisterKeyBind
+    private final KeyHolder lockSlotKeybind = new KeyHolder(
+            "Lock Slot", GLFW.GLFW_KEY_H, "Wynntils", true, ItemLockFeature::tryChangeLockStateOnHoveredSlot);
+
+    @Config(visible = false)
+    private static Map<Integer, Set<Integer>> classSlotLockMap = new HashMap<>();
+
+    @TypeOverride
+    private static final Type classSlotLockMapType = new TypeToken<HashMap<Integer, Set<Integer>>>() {}.getType();
+
+    @Config
+    public static boolean blockAllActionsOnLockedItems = false;
+
+    @SubscribeEvent
+    public void onKeyPress(InventoryKeyPressEvent e) {
+        if (lockSlotKeybind.getKeybind().matches(e.getKeyCode(), e.getScanCode())) {
+            lockSlotKeybind.onPress();
+        }
+    }
+
+    @SubscribeEvent
+    public void onContainerRender(ContainerRenderEvent event) {
+        if (!(event.getScreen() instanceof AbstractContainerScreen<?> abstractContainerScreen)) return;
+
+        Character character = WynnUtils.getCharacter();
+
+        if (character == null || character.getCharacterInfo() == null) return;
+
+        for (Integer slotId :
```

Comment:
I'm vaguely worried that we have a loop in render code. In all practical situations, this is unlikely to matter much, since there are only so many characters you can have. But as a general principle I really want to see rendering code being as minimal as possible. You could extract the inner set of locked slots into a cache, and update that cache whenever the player switches character. I'm pretty sure the Character model sends out an event when that happens, that you could listen to.